sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -213,20 +227,37 @@ data ProcessorMessage = Flush | Shutdown
 batchProcessor :: MonadIO m => BatchTimeoutConfig -> Exporter ImmutableSpan -> m Processor
 batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
   batch <- newIORef $ boundedMap maxQueueSize
-  workSignal <- newEmptyMVar
-  worker <- async $ loop $ do
-    req <-
-      liftIO $
-        timeout (millisToMicros scheduledDelayMillis) $
-          takeMVar workSignal
-    batchToProcess <- liftIO $ atomicModifyIORef' batch buildExport
-    res <- liftIO $ Exporter.exporterExport exporter batchToProcess
-
-    -- if we were asked to shutdown, quit cleanly after this batch
-    -- FIXME: this could lose batches if there's more than one in queue?
-    case req of
-      Just Shutdown -> throwE res
-      _ -> pure ()
+  workSignal <- newEmptyTMVarIO
+  shutdownSignal <- newEmptyTMVarIO
+  let workerAction = do
+        delay <- registerDelay (millisToMicros scheduledDelayMillis)
+        req <- atomically $ do
+          msum
+            [ const Flush <$> do
+                continue <- readTVar delay
+                check continue
+            , const Flush <$> takeTMVar workSignal
+            , const Shutdown <$> takeTMVar shutdownSignal
+            ]
```

Comment:
This is an async exception safe way to deal with `timeout` and reading from a variable. `registerDelay` forks a thread that merely writes to the `TVar Bool` when the elapsed time has completed. `takeTMVar` will block until the var is full, and then empty it.

Suppose we have a write to `Flush` and `Shutdown` at the same time. If the RTS picks `Flush`, it's no big deal - we run the HTTP action, then we loop again. The `Shutdown` var will be full, and will "flush" and then shutdown.